PlacesHistoryStorage crash in Sample Browser

This seems to be a regression. Sample browser is crashing quite frequently now when moving from background to foreground or loading a new page.

```
 java.lang.IllegalStateException: must call init first
        at mozilla.components.browser.storage.sync.RustPlacesConnection.writer(Connection.kt:95)
        at mozilla.components.browser.storage.sync.PlacesHistoryStorage$recordObservation$2.invokeSuspend(PlacesHistoryStorage.kt:63)
        at mozilla.components.browser.storage.sync.PlacesHistoryStorage$recordObservation$2.invoke(Unknown Source:10)
        at kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturn(Undispatched.kt:91)
        at kotlinx.coroutines.BuildersKt__Builders_commonKt.withContext(Builders.common.kt:154)
        at kotlinx.coroutines.BuildersKt.withContext(Unknown Source:1)
        at mozilla.components.browser.storage.sync.PlacesHistoryStorage.recordObservation$suspendImpl(PlacesHistoryStorage.kt:59)
        at mozilla.components.browser.storage.sync.PlacesHistoryStorage.recordObservation(Unknown Source:0)
        at mozilla.components.feature.session.HistoryDelegate.onTitleChanged(HistoryDelegate.kt:28)
        at mozilla.components.browser.engine.gecko.GeckoEngineSession$createContentDelegate$1$onTitleChange$$inlined$let$lambda$1.invokeSuspend(GeckoEngineSession.kt:700)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
```

Different trace for `onDestroy`:

```
    java.lang.RuntimeException: Unable to destroy activity {org.mozilla.samples.browser/org.mozilla.samples.browser.BrowserActivity}: java.lang.IllegalStateException: must call init first
        at android.app.ActivityThread.performDestroyActivity(ActivityThread.java:4941)
        at android.app.ActivityThread.handleDestroyActivity(ActivityThread.java:4970)
        at android.app.servertransaction.DestroyActivityItem.execute(DestroyActivityItem.java:44)
        at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2016)
        at android.os.Handler.dispatchMessage(Handler.java:107)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
     Caused by: java.lang.IllegalStateException: must call init first
        at mozilla.components.browser.storage.sync.RustPlacesConnection.close(Connection.kt:127)
        at mozilla.components.browser.storage.sync.PlacesStorage.cleanup(PlacesStorage.kt:64)
        at org.mozilla.samples.browser.BrowserActivity.onDestroy(BrowserActivity.kt:60)
        at android.app.Activity.performDestroy(Activity.java:8071)
```

--- a/components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/Connection.kt
+++ b/components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/Connection.kt
@@ -61,6 +61,7 @@ internal interface Connection : Closeable {
 /**
  * A singleton implementation of the [Connection] interface backed by the Rust Places library.
  */
+@Suppress("TooManyFunctions")
 internal object RustPlacesConnection : Connection {
     @GuardedBy("this")
     private var api: PlacesApi? = null
@@ -82,8 +83,9 @@ internal object RustPlacesConnection : Connection {
     }
 
     override fun getHandle(): Long {
+        val api = safeGetApi()
         check(api != null) { "must call init first" }
-        return api!!.getHandle()
+        return api.getHandle()
     }
 
     override fun reader(): PlacesReaderConnection = synchronized(this) {
@@ -92,40 +94,50 @@ internal object RustPlacesConnection : Connection {
     }
 
     override fun writer(): PlacesWriterConnection {
+        val api = safeGetApi()
         check(api != null) { "must call init first" }
-        return api!!.getWriter()
+        return api.getWriter()
     }
 
     override fun syncHistory(syncInfo: SyncAuthInfo) {
+        val api = safeGetApi()
         check(api != null) { "must call init first" }
-        val ping = api!!.syncHistory(syncInfo.into())
+        val ping = api.syncHistory(syncInfo.into())
         SyncTelemetry.processHistoryPing(ping)
     }
 
     override fun syncBookmarks(syncInfo: SyncAuthInfo) {
+        val api = safeGetApi()
         check(api != null) { "must call init first" }
-        val ping = api!!.syncBookmarks(syncInfo.into())
+        val ping = api.syncBookmarks(syncInfo.into())
         SyncTelemetry.processBookmarksPing(ping)
     }
 
     override fun importVisitsFromFennec(dbPath: String): JSONObject {
+        val api = safeGetApi()
         check(api != null) { "must call init first" }
-        return api!!.importVisitsFromFennec(dbPath)
+        return api.importVisitsFromFennec(dbPath)
     }
 
     override fun importBookmarksFromFennec(dbPath: String): JSONObject {
+        val api = safeGetApi()
         check(api != null) { "must call init first" }
-        return api!!.importBookmarksFromFennec(dbPath)
+        return api.importBookmarksFromFennec(dbPath)
     }
 
     override fun readPinnedSitesFromFennec(dbPath: String): List<BookmarkNode> {
+        val api = safeGetApi()
         check(api != null) { "must call init first" }
-        return api!!.importPinnedSitesFromFennec(dbPath).map { it.asBookmarkNode() }
+        return api.importPinnedSitesFromFennec(dbPath).map { it.asBookmarkNode() }
     }
 
     override fun close() = synchronized(this) {
         check(api != null) { "must call init first" }
         api!!.close()
         api = null
+    }
+
+    private fun safeGetApi(): PlacesApi? = synchronized(this) {
+        return this.api
     }
 }

--- a/samples/browser/src/main/java/org/mozilla/samples/browser/BrowserActivity.kt
+++ b/samples/browser/src/main/java/org/mozilla/samples/browser/BrowserActivity.kt
@@ -54,12 +54,6 @@ open class BrowserActivity : AppCompatActivity(), ComponentCallbacks2 {
         lifecycle.addObserver(webExtensionPopupFeature)
     }
 
-    override fun onDestroy() {
-        super.onDestroy()
-
-        components.historyStorage.cleanup()
-    }
-
     override fun onBackPressed() {
         supportFragmentManager.fragments.forEach {
             if (it is UserInteractionHandler && it.onBackPressed()) {

